### ISSUE TITLE
Bug: [prefer-nullish-coalescing] unexpected fails on condition with generics variable which can be `null` or `undefined`

### ISSUE THREAD

**Issue by zhangbenber** [typescript-eslint/typescript-eslint#7842]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygHgCoD4AUAbgIYA2AXItogD6LgAmKwMYKDAlIgN4BQiiAE4ooIQUhKlEAQgC8s%2BmCYs2DRAH5EkxJQDkwOHF0BuXgF9evVBnxgQpUh2NA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6MAG1IvaOOiQAuuDABfEPqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
function test<T>(val: T | undefined) {
  return val !== undefined ? val : 'foo';
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": ["error"]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect there's no errors reported.

### Actual Result

I get the following error:
```
Prefer using nullish coalescing operator (`??`) instead of a ternary expression, as it is simpler to read. 2:10 - 2:41
```
And I get a Fix to nullish coalescing operator (`??`), which changes the logic of the function.

### Additional Info

If I change the type of `val` to `T | undefined | null`, everything goes OK.

However, the generic `T` can also be `null`. The error does not disappear even if I make `T extends null`. In that situation, `val !== undefined ? val : 'foo'` is not equivalent to `val ?? 'foo'` and the fix may break the code if I call `test(null)`.

The same thing also happens in the following code:
```typescript
function test<T>(val: T | null) {
  return val !== null ? val : 'foo';
}
test(undefined);
```

**Comment by bradzacher** [comment]
In the most technical sense - yes, you are correct - `T` could be `null`, or it could be `U | null`.

It does seem pretty odd to me that any code written in this manner would be making an intentional and clear distinction between `null` and `undefined`. It would certainly be a niche edge-case.

Can you provide a little more context behind the code? Does the distinction matter?

**Comment by reilem** [comment]
I am getting false positives from this as well. Example:

```
let a: string | undefined = undefined;
let b: boolean = false;
let c: boolean = true;

// here some code that might modify a, b, c
// ...

if ((a && b) || c) {
    console.log("This is a test");
}
```
Will throw the error: `ESLint: Prefer using nullish coalescing operator ('??') instead of a logical or ('||'), as it is a safer operator.(@typescript-eslint/prefer-nullish-coalescing)`

### This is bad: replacing `||` with `??` will completely change the logic of this code.

Let's say `a = ""`, `b = false` and `c = true` then using the `||` we get:
1. `(a && b) -> ""`
2. `"" || true -> true`
3. The logic evaluates to a boolean true
4. The code will log "This is a test" to console

However if we follow the linter suggestion and use `??` we will get:

1. `(a && b) -> ""`
2. `"" ?? true -> ""`
3. An empty string is falsey
4. The code will not log anything

I know that there are some options I can probably configure to reduce the risk of this being flagged. However, I feel like this isn't optimal default behaviour. Less experienced developers could easily run into bugs using this rule.

**Comment by bradzacher** [comment]
@reilem https://typescript-eslint.io/rules/prefer-nullish-coalescing/#ignoreconditionaltests

**Comment by zhangbenber** [comment]
> Can you provide a little more context behind the code? Does the distinction matter?

@bradzacher Consider that I have a `Property` class, which stores a field value in a property set, and provide the getter and setter for the user. It also has a default value, which is indicated by `undefined` in value.

```typescript
class Property<T> {
  defaultValue: T;
  value: T | undefined; // undefined indicates default fallback

  constructor(defaultValue: T) {
    this.defaultValue = defaultValue;
  }

  set(value?: T | undefined) {
    this.value = value;
  }

  get() {
    // We focused on this line
    return this.value !== undefined ? this.value : this.defaultValue;
  }

  // ...
}
```

And we can use like this:

```typescript
const httpRequestProperties = {
  host: new Property('localhost'),
  // Property<string>, which must be set and cannot be null

  port: new Property(80),
  // Property<number>, which must be set and cannot be null

  username: new Property<string | null>('root'),
  // Can be null, and null means an empty value (anonymous)

  password: new Property<string | null>('p@ssword'),
  // Can be null, and null means an empty value (anonymous)

  // We should not specify a default value for usernames and passwords, and it's only for demonstration

  // ...
}
```

Notice that we explicitly defined the `string | null` type for the generics of `username` and `password` field. If we want an anonymous access, we can:

```typescript
httpRequestProperties.username.set(null);
httpRequestProperties.password.set(null);
```

and this will override their default values, and return `null` when `get()` method is called.

However, the code above gets the error:

```
Prefer using nullish coalescing operator (`??`) instead of a ternary expression, as it is simpler to read. 14:12 - 14:69
```

And if I accept the fix and change to:

```typescript
  get() {
    return this.value ?? this.defaultValue;
  }
```

you may find the logic is broken. I will get `'root'` when I call `httpRequestProperties.username.get()` even if it is set to `null`.

An attempt to solve this is to redefine the type of `value` to `T | undefined | null`, and now typescript-eslint is happy. But it changes the signature of the `get()` method, especially for these non-nullish fields.

```typescript
class Property<T> {
  defaultValue: T;
  value: T | undefined | null; // We add the explicit `null`
  // ...

  get() {
    return this.value !== undefined ? this.value : this.defaultValue;
  }
}

const httpRequestProperties = {
  host: new Property('localhost'),
  // Property<string>, which must be set and cannot be null
  // ...
}

const host = httpRequestProperties.host.get();
typeof host // => `string | null`, but it should never be `null` here
```

And I must cast them into the right type manually, in the ungraceful manner.

In this situation to my view, `null` is part of the generic `T` and should be explicitly defined by the user. It is related to the user's usage and has the indication for the possibility of this generic being `null`. Defining `T | null` may break this indication, so that we get to the wrong way.

I will be happy if you want to fix this, or give any suggestions for this situation. Thank you!

**Comment by bradzacher** [comment]
*personally* I've found that adding meaning to the distinction between `null` vs `undefined` is generally a code smell given how easily `undefined` can creep in accidentally. If I were to design it I'd instead use a symbol to make it explicit that a value is supposed to be unset so that you can never accidentally fall into the default value case.

EG:
```ts
const EMPTY = Symbol();
class Property<T> {
  static readonly EMPTY = EMPTY;
  defaultValue: T;
  value: T | null | typeof EMPTY = EMPTY;

  set(value: T | typeof EMPTY) {
    this.value = value;
  }
  get(): T {
    return this.value === EMPTY ? this.defaultValue : this.value;
  }
}
```

That way you can also support cases like `new Property<string | undefined>()` without anything breaking unintentionally.

**Comment by zhangbenber** [comment]
Yes, I agree with you and that was what I exactly did to solve this. It's the perfect solution to this situation.

But this may have other shortcomings, not easy for serializing, for example. And I personally do not think distinction between `null` and `undefined` is unacceptable, since TypeScript does distinguish them (comparing to `undefined` vs `void`). JSON is another example, `null` and absence is different.

Or, could you report another type of rule violate for this case, or add some options to ignore this, as what the `ignorePrimitives` does? In this way one can prevent the dangerous fix action, especially for these auto-fixers.

**Comment by JoshuaKGoldberg** [comment]
Although I agree with Brad's points that this is some funky behavior, I also agree that it's technically a bug in the rule 😄. Accepting PRs.

Let's have the rule give a more specific error message because I think a lot of people get tripped up on type parameters being allowed to be `null`. TypeScript added a similar special casing in https://github.com/microsoft/TypeScript/pull/30394.

**Comment by ulrichstark** [comment]
I've noticed a similar issue that's #7055 but with a generic function.
If the logical or operator is applied on two generic and potentially null function parameters, it will show a false positive if the function is called with boolean as type for generic.

Reproduction Sandbox:
https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=MYewdgzgLgBMCuAnRBTMUBqBDANvFAXDAEYgg4pZgwA%2BMY8OOMAvDAGa4QoDcAsAChQkWAAdUANwCWIeBGx5CJMhSq16jZmyiJ8-IeGgxpEKcQoL8rGAHMUmKafMpLKABQJkaTLnwAaGHEUaVl5XxQASh4YAHoYmABRAA9RFGAoFAATIhMzC3DrHXxBQXZ4MHSZajsHJ3zFAB4AFQA%2BDyRUdFciJvUGJgCgkLlumF66fpwInr7NGABvQRgYVCgkak9On0VaOiGZEfDouJgAflOYAHdZHEySFCvEcBsYAAsUVHvgLDkHqHfqLlnK5rJwcNxBABfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Do we need a flag like `ignoreGenerics` or is this a bug?

**Comment by lorenzpfei** [comment]
Having the same issue using 
```typescript
const isDisabled = Boolean(
    priceIdLoading === price.priceId ||
      isPurchased ||
      (type === 'subscription' &&
        subscription &&
        !canUpgrade &&
        !isCurrentPlan),
  );
  ```
  
  when I change it to
  
  ```typescript
  const isDisabled = Boolean(
    priceIdLoading === price.priceId ||
      isPurchased ?? // <------
      (type === 'subscription' &&
        subscription &&
        !canUpgrade &&
        !isCurrentPlan),
  );
  ```
  it's not the same logic anymore

**Comment by bradzacher** [comment]
@lorenzpfei please don't post random code snippets with no context. How can we prove assistance without any types to understand your code? Always provide a playground link so people can quickly reproduce your code.

**Comment by OlivierZal** [comment]
@bradzacher, @JoshuaKGoldberg, @kirkwaiblinger,

what behaviour do we want for the following:

```ts
function test<T>(val: T | null) {
  return val ? val : 'foo';
}
```

**Comment by pfeileon** [comment]
This is a code snippet which everybody should understand just by reading it:
```
const x = '';
const y = '_';

// ESLint: Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`),
//         as it is a safer operator. (@typescript-eslint/prefer-nullish-coalescing)
const z = x || y; // z === '_'
const z = x ?? y; // z === ''
```
This rule is potentially harmful. That's a massive problem.
Imho the docs should be updated to make this clear and the rule should be cut from all predefined configs as long as it's not fixed.

An understandable rule would warn you that using the logical `||` operator in this way is potentially misleading and that you should be explicit:
```
const z = x === '' ? y : (x ?? y);
```

**Comment by bradzacher** [comment]
@pfeileon that is completely unrelated to this issue. Please do not add unrelated comments to issues as per our [contributing guide](https://typescript-eslint.io/contributing/issues#commenting)
